### ISSUE TITLE
 Could not load file or assembly 'System.Diagnostics.DiagnosticSource, Version=4.0.0.0

### ISSUE THREAD

**Issue by dslagman** [canton7/RestEase#34]
I reference RestEase from a net461 project. It builds fine but gives a runtime exception:

Message: System.IO.FileLoadException : Could not load file or assembly 'System.Diagnostics.DiagnosticSource, Version=4.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51' or one of its dependencies. The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040)

Any idea on how to solve this?

**Comment by canton7** [comment]
More details please:

1. Visual Studio version?
2. Is this a .net core project targetting net461, or what?
3. Full stack trace?

A quick google for "The located assembly's manifest definition does not match the assembly reference System.Diagnostics.DiagnosticSource" suggests it's not uncommon and has a couple of potential solutions, but it's hard to know which ones to narrow the search down to without more info.

**Comment by dslagman** [comment]
I'm using VS 2017. I'm using a net461 library that references RestEase version 1.3.3. 

However when I open a fresh solution with a fresh library targeting net461. There is on issue. However when I also start referencing the ABP nuget package (where doing that in my solution where I encountered the error as well) I know get the same error only now for Newtonsoft.Json, Version=6.0.0.0. 

Result StackTrace:	
at RestEase.RestClient.For[T]()
   at RestEaseFrom461.TimesheetApi_Tests.<Test_Get_Worklogs>d__0.MoveNext() in C:\workspace-samples\RestEaseFrom461\RestEaseFrom461Test\TimesheetApi_Tests.cs:line 15
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
Result Message:	System.IO.FileLoadException : Could not load file or assembly 'Newtonsoft.Json, Version=6.0.0.0, Culture=neutral, PublicKeyToken=30ad4fe6b2a6aeed' or one of its dependencies. The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040)

**Comment by canton7** [comment]
> I'm using a net461 library that references RestEase version 1.3.3.

Which net461 library is this? One that you wrote, or something else?

A google for "could not load file or assembly 'Newtonsoft.Json, Version=6.0.0.0, Culture=neutral, PublicKeyToken=30ad4fe6b2a6aeed' or one of its dependencies. The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040)" brings up lots of relevant results. Among them:

1. http://stackoverflow.com/questions/22685530/could-not-load-file-or-assembly-newtonsoft-json-or-one-of-its-dependencies-ma
2. http://stackoverflow.com/questions/17776090/newtonsoft-json-assembly-conflict

Basically, this is due to RestEase referencing an oldish version of Newtonsoft.Json (which is what it should do). Another NuGet package you added depends on a newer version of Newtonsoft.Json (also fine). Nuget looks at both constraints and installs the newer version. However when RestEase tries to load Newtonsoft.Json, it finds the older version and complains.

Now NuGet should have set up binding redirects in your app.config to tell the runtime that it's OK to find a newer version of Newtonsoft.Json when it expected an older one. I'm not sure why it hasn't done this, but those google links describe what to do fairly well.

**Comment by dslagman** [comment]
It is my own library. Referencing both RestEase and the ABP package. The ABP package indeed references the latest Netwonsoft.Json package. I'm still confused a bit which version of NewtonSoft.Json or System.Diagnostics.DiagnosticSource should be installed then. Or should there be two versions (the old, and the new one)?

**Comment by canton7** [comment]
> It is my own library.

Ah that might be why NuGet hasn't set up the binding redirects.

>  I'm still confused a bit which version of NewtonSoft.Json or System.Diagnostics.DiagnosticSource should be installed then. Or should there be two versions (the old, and the new one)?

You should install the latest versions, and set up binding redirects to tell the runtime that it's OK to load the latest version when an earlier version is expected.

**Comment by dslagman** [comment]
Great I've got it working! I was struggeling a bit with picking the right versions but the below app.config content did the trick. Thanks for your help!

```
<?xml version="1.0" encoding="utf-8" ?>
<configuration>
  <runtime>
    <assemblyBinding  xmlns="urn:schemas-microsoft-com:asm.v1">
      <dependentAssembly>
        <assemblyIdentity name="Newtonsoft.Json" publicKeyToken="30ad4fe6b2a6aeed"/>
        <bindingRedirect oldVersion="0.0.0.0-6.0.0.1" newVersion="9.0.0.0"/>
      </dependentAssembly>
    </assemblyBinding>
  </runtime>
</configuration>
```

**Comment by canton7** [comment]
Glad you got it working!

Apparently there's a NuGet powershell cmdlet which can (apparently) set the binding redirects up for you.

**Comment by dslagman** [comment]
I found a great tool by the way to scan a binaries directory for all references called asmspy.
